extmod/modure.c

```diff
@@ -372,7 +372,11 @@ STATIC mp_obj_t re_sub_helper(size_t n_args, const mp_obj_t *args) {
     // Add post-match string
     vstr_add_strn(&vstr_return, subj.begin, subj.end - subj.begin);
 
-    return mp_obj_new_str_from_vstr(mp_obj_get_type(where), &vstr_return);
+    if (mp_obj_get_type(where) == &mp_type_str) {
+        return mp_obj_new_str_from_vstr(&vstr_return);
+    } else {
+        return mp_obj_new_bytes_from_vstr(&vstr_return);
+    }
```

Comment:
Would be interesting to find a test case where doing a re match does raise a unicode error, and then check how CPython handles it.

Also, to reduce code size, maybe a helper like `mp_obj_new_str_type_from_vstr_checked(type, vstr)` could be added?